lightly/models/simsiam.py

```diff
@@ -0,0 +1,159 @@
+    and output dimension of 2048
+
+    Args:
+        in_dims:
+            Input dimension of the first linear layer.
+        h_dims: 
+            Hidden dimension of all the fully connected layers (should be a
+            bottleneck!)
+        out_dims: 
+            Output Dimension of the final linear layer.
+
+    Returns:
+        nn.Sequential:
+            The projection head.
+    """
+    l1 = nn.Sequential(nn.Linear(in_dims, h_dims),
+                       nn.BatchNorm1d(h_dims),
+                       nn.ReLU(inplace=True))
+
+    l2 = nn.Linear(h_dims, out_dims)
+
+    prediction = nn.Sequential(l1, l2)
+    return prediction
+
+
+def _projection_mlp(in_dims: int = 2048, 
+                    h_dims: int = 2048, 
+                    out_dims: int = 2048) -> nn.Sequential:
+    """Projection MLP. The original paper's implementation has 3 layers, with 
+    BN applied to its hidden fc layers but no ReLU on the output fc layer.
+
+    Args:
+        in_dims:
+            Input dimension of the first linear layer.
+        h_dims: 
+            Hidden dimension of all the fully connected layers.
+        out_dims: 
+            Output Dimension of the final linear layer.
+
+    Returns:
+        nn.Sequential:
+            The projection head.
+    """
+    l1 = nn.Sequential(nn.Linear(in_dims, h_dims),
+                       nn.BatchNorm1d(h_dims),
+                       nn.ReLU(inplace=True))
+
+    l2 = nn.Sequential(nn.Linear(h_dims, h_dims),
+                       nn.BatchNorm1d(h_dims),
+                       nn.ReLU(inplace=True))
+
+    l3 = nn.Sequential(nn.Linear(h_dims, out_dims),
+                       nn.BatchNorm1d(out_dims))
+
+    projection = nn.Sequential(l1, l2, l3)
+    return projection
+
+
+class SimSiam(nn.Module, _StateDictLoaderMixin):
+    """ Implementation of SimSiam network
+
+    Attributes:
+        backbone:
+            TODO
+        width:
+            Width of the ResNet.
+        num_ftrs:
+            Dimension of the embedding (before the projection head).
+        proj_hidden_dim:
+            TODO
+        pred_hidden_dim:
+            TODO
+        out_dim:
+            Dimension of the output (after the projection head).
+
+    """
+
+    def __init__(self,
+                 backbone: nn.Module,
+                 num_ftrs: int = 2048,
+                 proj_hidden_dim: int = 2048,
+                 pred_hidden_dim: int = 512,
+                 out_dim: int = 2048):
+
+        super(SimSiam, self).__init__()
+
+        self.backbone = backbone
+        self.num_ftrs = num_ftrs
+        self.proj_hidden_dim = proj_hidden_dim
+        self.pred_hidden_dim = pred_hidden_dim
+        self.out_dim = out_dim
+
+        self.projection_mlp = \
+            _projection_mlp(num_ftrs, proj_hidden_dim, out_dim)
+
+        self.prediction_mlp = \
+            _prediction_mlp(num_ftrs, pred_hidden_dim, out_dim)
+        
```

Comment:
This should be 
```
self.prediction_mlp = \
    _prediction_mlp(out_dim, pred_hidden_dim, out_dim)
```
Because then I can change the `num_ftrs` without breaking the forward pass, right?

Reply:
If anything, would it not be

```Python
self.prediction_mlp = _prediction_mlp(num_ftrs, pred_hidden_dim, num_ftrs)
```

Not totally sure what you mean by changing `num_ftrs` without breaking the forward pass?

Reply:
Currently, if `num_ftrs != out_dim`, the forward pass breaks because the output of the projection head has a different shape than the input of the prediction head. Therefore, my suggestion was to change the input dimension of the prediction head to the output dimension of the projection head.

Reply:
Ah gotcha, yes you're right. I still mix up prediction and projection in my head 😀